File: DandDy/Models/Character.swift
import Foundation

enum Alignment: String, Codable, CaseIterable {
    case lawfulGood = "lawful_good"
    case neutralGood = "neutral_good"
    case chaoticGood = "chaotic_good"
    case lawfulNeutral = "lawful_neutral"
    case trueNeutral = "true_neutral"
    case chaoticNeutral = "chaotic_neutral"
    case lawfulEvil = "lawful_evil"
    case neutralEvil = "neutral_evil"
    case chaoticEvil = "chaotic_evil"
    
    var displayName: String {
        switch self {
        case .lawfulGood: return "Lawful Good"
        case .neutralGood: return "Neutral Good"
        case .chaoticGood: return "Chaotic Good"
        case .lawfulNeutral: return "Lawful Neutral"
        case .trueNeutral: return "True Neutral"
        case .chaoticNeutral: return "Chaotic Neutral"
        case .lawfulEvil: return "Lawful Evil"
        case .neutralEvil: return "Neutral Evil"
        case .chaoticEvil: return "Chaotic Evil"
        }
    }
}

struct Character: Codable, Identifiable {
    let id: Int
    let ownerId: Int
    let campaignId: Int?
    
    // Basic Info
    var name: String
    var race: String
    var characterClass: String
    var level: Int
    var background: String?
    var alignment: Alignment?
    var experiencePoints: Int
    
    // Ability Scores
    var strength: Int
    var dexterity: Int
    var constitution: Int
    var intelligence: Int
    var wisdom: Int
    var charisma: Int
    
    // Combat Stats
    var hitPointsMax: Int
    var hitPointsCurrent: Int
    var hitPointsTemp: Int
    var armorClass: Int
    var initiative: Int
    var speed: Int
    
    // Death Saves
    var deathSaveSuccesses: Int
    var deathSaveFailures: Int
    
    // Proficiencies
    var savingThrowProficiencies: [String]
    var skillProficiencies: [String]
    var skillExpertises: [String]
    
    // Features
    var racialTraits: [[String: String]]
    var classFeatures: [[String: String]]
    var feats: [[String: String]]
    
    // Personality
    var personalityTraits: String?
    var ideals: String?
    var bonds: String?
    var flaws: String?
    
    // Appearance
    var appearance: String?
    var backstory: String?
    
    // Inventory
    var inventory: [[String: String]]
    
    // Spellcasting
    var spellcastingAbility: String?
    var spellSaveDc: Int?
    var spellAttackBonus: Int?
    var spellSlots: [String: Int]
    var spellSlotsUsed: [String: Int]
    var spellsKnown: [[String: String]]
    var spellsPrepared: [String]
    
    // Combat
    var conditions: [String]
    var attacks: [[String: String]]
    
    // Currency
    var copperPieces: Int
    var silverPieces: Int
    var electrumPieces: Int
    var goldPieces: Int
    var platinumPieces: Int
    
    enum CodingKeys: String, CodingKey {
        case id, name, race, level, background, alignment
        case ownerId = "owner_id"
        case campaignId = "campaign_id"
        case characterClass = "character_class"
        case experiencePoints = "experience_points"
        case strength, dexterity, constitution, intelligence, wisdom, charisma
        case hitPointsMax = "hit_points_max"
        case hitPointsCurrent = "hit_points_current"
        case hitPointsTemp = "hit_points_temp"
        case armorClass = "armor_class"
        case initiative, speed
        case deathSaveSuccesses = "death_save_successes"
        case deathSaveFailures = "death_save_failures"
        case savingThrowProficiencies = "saving_throw_proficiencies"
        case skillProficiencies = "skill_proficiencies"
        case skillExpertises = "skill_expertises"
        case racialTraits = "racial_traits"
        case classFeatures = "class_features"
        case feats
        case personalityTraits = "personality_traits"
        case ideals, bonds, flaws
        case appearance, backstory
        case inventory
        case spellcastingAbility = "spellcasting_ability"
        case spellSaveDc = "spell_save_dc"
        case spellAttackBonus = "spell_attack_bonus"
        case spellSlots = "spell_slots"
        case spellSlotsUsed = "spell_slots_used"
        case spellsKnown = "spells_known"
        case spellsPrepared = "spells_prepared"
        case conditions, attacks
        case copperPieces = "copper_pieces"
        case silverPieces = "silver_pieces"
        case electrumPieces = "electrum_pieces"
        case goldPieces = "gold_pieces"
        case platinumPieces = "platinum_pieces"
    }
}

struct CharacterCreate: Codable {
    var name: String
    var race: String
    var characterClass: String
    var level: Int = 1
    var background: String?
    var alignment: Alignment?
    var experiencePoints: Int = 0
    
    var strength: Int
    var dexterity: Int
    var constitution: Int
    var intelligence: Int
    var wisdom: Int
    var charisma: Int
    
    var hitPointsMax: Int
    var hitPointsCurrent: Int
    var hitPointsTemp: Int = 0
    var armorClass: Int
    var initiative: Int = 0
    var speed: Int = 30
    
    var deathSaveSuccesses: Int = 0
    var deathSaveFailures: Int = 0
    
    var savingThrowProficiencies: [String] = []
    var skillProficiencies: [String] = []
    var skillExpertises: [String] = []
    
    var racialTraits: [[String: String]] = []
    var classFeatures: [[String: String]] = []
    var feats: [[String: String]] = []
    
    var personalityTraits: String?
    var ideals: String?
    var bonds: String?
    var flaws: String?
    
    var appearance: String?
    var backstory: String?
    
    var inventory: [[String: String]] = []
    
    var spellcastingAbility: String?
    var spellSaveDc: Int?
    var spellAttackBonus: Int?
    var spellSlots: [String: Int] = [:]
    var spellSlotsUsed: [String: Int] = [:]
    var spellsKnown: [[String: String]] = []
    var spellsPrepared: [String] = []
    
    var conditions: [String] = []
    var attacks: [[String: String]] = []
    
    var copperPieces: Int = 0
    var silverPieces: Int = 0
    var electrumPieces: Int = 0
    var goldPieces: Int = 0
    var platinumPieces: Int = 0
    
    var campaignId: Int?
    
    enum CodingKeys: String, CodingKey {
        case name, race, level, background, alignment
        case characterClass = "character_class"
        case experiencePoints = "experience_points"
        case strength, dexterity, constitution, intelligence, wisdom, charisma
        case hitPointsMax = "hit_points_max"
        case hitPointsCurrent = "hit_points_current"
        case hitPointsTemp = "hit_points_temp"
        case armorClass = "armor_class"
        case initiative, speed
        case deathSaveSuccesses = "death_save_successes"
        case deathSaveFailures = "death_save_failures"
        case savingThrowProficiencies = "saving_throw_proficiencies"
        case skillProficiencies = "skill_proficiencies"
        case skillExpertises = "skill_expertises"
        case racialTraits = "racial_traits"
        case classFeatures = "class_features"
        case feats
        case personalityTraits = "personality_traits"
        case ideals, bonds, flaws
        case appearance, backstory
        case inventory
        case spellcastingAbility = "spellcasting_ability"
        case spellSaveDc = "spell_save_dc"
        case spellAttackBonus = "spell_attack_bonus"
        case spellSlots = "spell_slots"
        case spellSlotsUsed = "spell_slots_used"
        case spellsKnown = "spells_known"
        case spellsPrepared = "spells_prepared"
        case conditions, attacks
        case copperPieces = "copper_pieces"
        case silverPieces = "silver_pieces"
        case electrumPieces = "electrum_pieces"
        case goldPieces = "gold_pieces"
        case platinumPieces = "platinum_pieces"
        case campaignId = "campaign_id"
    }
}

File: backend/models/character.py
from sqlalchemy import Column, Integer, String, ForeignKey, JSON, Enum, DateTime, Index, Boolean
from sqlalchemy.orm import relationship
import enum
from datetime import datetime
from database.database import Base

class Alignment(enum.Enum):
    LAWFUL_GOOD = "lawful_good"
    NEUTRAL_GOOD = "neutral_good"
    CHAOTIC_GOOD = "chaotic_good"
    LAWFUL_NEUTRAL = "lawful_neutral"
    TRUE_NEUTRAL = "true_neutral"
    CHAOTIC_NEUTRAL = "chaotic_neutral"
    LAWFUL_EVIL = "lawful_evil"
    NEUTRAL_EVIL = "neutral_evil"
    CHAOTIC_EVIL = "chaotic_evil"

class Sex(enum.Enum):
    MALE = "male"
    FEMALE = "female"

class Character(Base):
    __tablename__ = "characters"
    __table_args__ = (
        # Indexes to speed up common queries:
        # - owner_id: listing a user's characters
        # - campaign_id: loading characters in a campaign
        # - updated_at: sorting/filtering by last modified
        Index("idx_characters_owner_id", "owner_id"),
        Index("idx_characters_campaign_id", "campaign_id"),
        Index("idx_characters_updated_at", "updated_at"),
    )
    
    id = Column(Integer, primary_key=True, index=True)
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    campaign_id = Column(Integer, ForeignKey("campaigns.id"), nullable=True)
    
    # Basic Info
    name = Column(String, nullable=False)
    race = Column(String, nullable=False)
    character_class = Column(String, nullable=False)
    level = Column(Integer, default=1, nullable=False)
    background = Column(String, nullable=True)
    alignment = Column(Enum(Alignment), nullable=True)
    sex = Column(Enum(Sex), nullable=True)
    experience_points = Column(Integer, default=0, nullable=False)
    
    # Ability Scores
    strength = Column(Integer, nullable=False)
    dexterity = Column(Integer, nullable=False)
    constitution = Column(Integer, nullable=False)
    intelligence = Column(Integer, nullable=False)
    wisdom = Column(Integer, nullable=False)
    charisma = Column(Integer, nullable=False)
    
    # Combat Stats
    hit_points_max = Column(Integer, nullable=False)
    hit_points_current = Column(Integer, nullable=False)
    hit_points_temp = Column(Integer, default=0, nullable=False)
    armor_class = Column(Integer, nullable=False)
    initiative = Column(Integer, default=0, nullable=False)
    speed = Column(Integer, default=30, nullable=False)
    
    # NOTE: hit_dice_current and class_resources columns planned but not yet migrated.
    # Uncomment when database migration completes:
    # hit_dice_current = Column(Integer, nullable=True)  # None means full (equals level)
    # class_resources = Column(JSON, default=dict, nullable=False)  # Ki, Rage, etc.

    # Death Saves
    death_save_successes = Column(Integer, default=0, nullable=False)
    death_save_failures = Column(Integer, default=0, nullable=False)
    
    # Proficiencies (stored as JSON arrays)
    saving_throw_proficiencies = Column(JSON, default=list, nullable=False)  # ["str", "con"]
    skill_proficiencies = Column(JSON, default=list, nullable=False)  # ["athletics", "perception"]
    skill_expertises = Column(JSON, default=list, nullable=False)  # ["stealth"]
    tool_proficiencies = Column(JSON, default=list, nullable=False)  # ["thieves-tools", "gaming-set"]
    languages = Column(JSON, default=list, nullable=False)  # ["Common", "Elvish", "Dwarvish"]
    
    # Features and Traits
    racial_traits = Column(JSON, default=list, nullable=False)
    class_features = Column(JSON, default=list, nullable=False)
    feats = Column(JSON, default=list, nullable=False)
    background_feature = Column(JSON, default=dict, nullable=False)  # {"name": "...", "description": "..."}
    
    # Personality
    personality_traits = Column(String, nullable=True)
    ideals = Column(String, nullable=True)
    bonds = Column(String, nullable=True)
    flaws = Column(String, nullable=True)
    
    # Appearance
    appearance = Column(String, nullable=True)
    backstory = Column(String, nullable=True)
    
    # Portrait Data
    ascii_portrait = Column(String, nullable=True)  # ASCII art portrait (text)
    original_portrait_url = Column(String, nullable=True)  # URL to generated image
    custom_portrait_ascii = Column(String, nullable=True)  # Custom AI-generated ASCII
    custom_portrait_count = Column(Integer, default=0, nullable=False)  # Number of custom portraits generated
    portrait_metadata = Column(JSON, default=dict, nullable=False)  # Additional portrait info (key, source, etc)
    
    # Inventory (JSON array of items)
    inventory = Column(JSON, default=list, nullable=False)
    
    # Spellcasting
    spellcasting_ability = Column(String, nullable=True)  # "int", "wis", "cha"
    spell_save_dc = Column(Integer, nullable=True)
    spell_attack_bonus = Column(Integer, nullable=True)
    spell_slots = Column(JSON, default=dict, nullable=False)  # {"1": 4, "2": 3, "3": 2}
    spell_slots_used = Column(JSON, default=dict, nullable=False)  # {"1": 2, "2": 1, "3": 0}
    cantrips = Column(JSON, default=list, nullable=False)  # ["Fire Bolt", "Mage Hand", ...]
    spells_known = Column(JSON, default=list, nullable=False)  # ["Fireball", "Shield", ...]
    spells_prepared = Column(JSON, default=list, nullable=False)  # ["Fireball", "Shield", ...]
    
    # Combat Conditions
    conditions = Column(JSON, default=list, nullable=False)  # ["prone", "blinded"]
    
    # Attacks and Actions
    attacks = Column(JSON, default=list, nullable=False)  # [{"name": "Longsword", "bonus": 5, "damage": "1d8+3"}]
    
    # Currency
    copper_pieces = Column(Integer, default=0, nullable=False)
    silver_pieces = Column(Integer, default=0, nullable=False)
    electrum_pieces = Column(Integer, default=0, nullable=False)
    gold_pieces = Column(Integer, default=0, nullable=False)
    platinum_pieces = Column(Integer, default=0, nullable=False)

    # Demo Mode
    is_demo = Column(Boolean, default=False, nullable=False)  # If true, shown in demo mode to non-logged-in users

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Track who last updated (for shared characters)
    last_updated_by_id = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    
    # Relationships
    owner = relationship("User", back_populates="characters", foreign_keys=[owner_id])
    last_updated_by = relationship("User", foreign_keys=[last_updated_by_id])
    campaign = relationship("Campaign", back_populates="characters")
    collaborators = relationship("CharacterCollaborator", back_populates="character", cascade="all, delete-orphan")
    
    # Campaign tracking relationships
    campaign_membership = relationship("CampaignMember", back_populates="character", uselist=False)
    sessions = relationship("Session", back_populates="character", cascade="all, delete-orphan")
    session_logs = relationship("SessionLog", back_populates="character", cascade="all, delete-orphan")
    
    # Journal relationships
    journal_entries = relationship("JournalEntry", back_populates="character", cascade="all, delete-orphan")
    character_updates = relationship("CharacterUpdate", back_populates="character", cascade="all, delete-orphan")


